### PR TITLE
Add methods for converting lockfile back to string

### PR DESCRIPTION
**Pull request by chmeliik**
Add to_file, to_str as counterparts to from_file, from_str

Signed-off-by: Adam Cmiel <acmiel@redhat.com>

### CODE REVIEW

**Comment by chmeliik** [comment]
CC @nirzari @ejegrova (can't add you as reviewers)

**Comment by chmeliik** [comment]
Will be useful for CLOUDBLD-3314

File: pyarn/lexer.py [hunk 1]
```diff
@@ -35,6 +35,11 @@ def t_NUMBER(t):
     return t
 
 
+# Unquoted string regex (see t_STRING)
+# Docstrings cannot use string interpolation, this value is used by other modules
```
Comment (nirzari) [inline pyarn/lexer.py:39]: Nitpick
```suggestion
# Docstrings can not use string interpolation, this value is used by other modules
```

File: pyarn/lexer.py [hunk 2]
```diff
@@ -35,6 +35,11 @@ def t_NUMBER(t):
...
+UNQUOTED_STRING = r'[a-zA-Z/.-]([^\s\n,]*[^\s\n,:])?'
```
Comment (nirzari) [inline pyarn/lexer.py:40]: Just out of curiosity, doesn't black prefer double quotes over single?

**Comment by nirzari** [comment]
Semantically it makes sense to me. I'd wait for one more approval but LGTM :+1:

File: pyarn/lexer.py (hunk 2 above)
Comment (chmeliik) [reply pyarn/lexer.py:40]: It does, but this project does not (yet?) use `black` and most of the strings I've seen here use single quotes

File: pyarn/lexer.py (hunk 1 above)
Comment (chmeliik) [reply pyarn/lexer.py:39]: I double-checked some sources, e.g. https://english.stackexchange.com/questions/4510/why-is-cannot-spelled-as-one-word
TLDR both "cannot" and "can not" are acceptable, but "cannot" is more common

Comment (nirzari) [reply pyarn/lexer.py:39]: lol _The modern way of writing can not_ , perhaps I'm old

File: pyarn/lexer.py (hunk 2 above)
Comment (nirzari) [reply pyarn/lexer.py:40]: Yep. that's why question out of curiosity. Perhaps in the future we could just start using it which will fix it all.

**Review by nirzari — APPROVED** [review]

File: pyarn/lockfile.py [hunk 3]
```diff
@@ -88,3 +91,50 @@ def from_str(cls, lockfile_str):
...
+            self._dump(lockfile)
+
+    def to_str(self):
+        buffer = io.StringIO()
+        self._dump(buffer)
+        return buffer.getvalue()
+
+    def _dump(self, outfile):
+        # Does not preserve any comments, but this one is required
+        outfile.write('# yarn lockfile v1\n')
```
Comment (athos-ribeiro) [inline pyarn/lockfile.py]: Should this be a constant somewhere?

I realized later after processing this string that it is only used for lockfile v1.

File: pyarn/lexer.py (hunk 2 above)
Comment (athos-ribeiro) [reply pyarn/lexer.py:40]: I did not bring black in for the first commits intentionally. I thought pyarn was quite small and I did not see any need to enforce styles in the sense black does it. +1 to bring it in if you feel it would be worth it.

Comment (athos-ribeiro) [reply pyarn/lexer.py:40]: Why is this defined in the lexer and not in `lockfile.py` itself? ?

File: tests/test_lockfile.py [hunk 4]
```diff
@@ -165,3 +166,59 @@ def test_package_with_comma():
...
+    lockfile.Lockfile('1', DATA_TO_DUMP).to_file(output_path)
+    assert output_path.read_text() == EXPECTED_CONTENT
+
+
+def test_to_str():
+    content = lockfile.Lockfile('1', DATA_TO_DUMP).to_str()
+    assert content == EXPECTED_CONTENT
+
+
+def test_roundtrip(all_test_files, tmp_path):
```
Comment (athos-ribeiro) [inline tests/test_lockfile.py:218]: I like the test name! :rofl:

File: tests/test_lockfile.py [hunk 5]
```diff
@@ -165,3 +166,59 @@ def test_package_with_comma():
...
+    for test_file in all_test_files:
+        out_file = tmp_path / os.path.basename(test_file)
+        # Check that load -> dump -> load produces expected results
```
Comment (athos-ribeiro) [inline tests/test_lockfile.py:221]: nice!

**Review by athos-ribeiro — COMMENTED** [review]
LGTM

Just a couple questions/nitpicks inline

File: pyarn/lexer.py (hunk 2 above)
Comment (chmeliik) [reply pyarn/lexer.py:40]: > Why is this defined in the lexer and not in `lockfile.py` itself? ?

I wanted this to be defined as close to `t_STRING` as possible in the hope that if someone changes the relevant part of the `t_STRING` regex, they will remember to update the `UNQUOTED_STRING` regex accordingly (it is just copy-pasted from there)

File: pyarn/lockfile.py (hunk 3 above)
Comment (chmeliik) [reply pyarn/lockfile.py]: Should the constant also be used when determining the file version instead of the regex?

File: pyarn/lexer.py (hunk 2 above)
Comment (athos-ribeiro) [reply pyarn/lexer.py:40]: Sounds good!

File: pyarn/lockfile.py (hunk 3 above)
Comment (athos-ribeiro) [reply pyarn/lockfile.py]: Sounds good, yes!

Comment (chmeliik) [reply pyarn/lockfile.py]: ok, changed :-)

**Review by athos-ribeiro — APPROVED** [review]
